src/VisualStudio/IntegrationTests/AbstractEditorTests.cs

```diff
@@ -0,0 +1,231 @@
+        {
+            _visualStudioWorkspace.UseSuggestionMode = true;
+        }
+
+        protected void InvokeCompletionList()
+        {
+            ExecuteCommand(WellKnownCommandNames.ListMembers);
+            WaitForAsyncOperations(FeatureAttribute.CompletionSet);
+        }
+
+        protected void ExecuteCommand(string commandName)
+        {
+            _visualStudio.Instance.ExecuteCommand(commandName);
+        }
+
+        protected void VerifyCurrentLineText(string expectedText, bool trimWhitespace = true)
+        {
+            var caretStartIndex = expectedText.IndexOf("$$");
+
+            if (caretStartIndex >= 0)
+            {
+                var caretEndIndex = caretStartIndex + "$$".Length;
+
+                var expectedTextBeforeCaret = caretStartIndex < expectedText.Length
+                    ? expectedText.Substring(0, caretStartIndex)
+                    : expectedText;
+
+                var expectedTextAfterCaret = caretEndIndex < expectedText.Length
+                    ? expectedText.Substring(caretEndIndex)
+                    : string.Empty;
+
+                var lineText = _editor.GetCurrentLineText();
+
+                if (trimWhitespace)
+                {
+                    if (caretStartIndex == 0)
+                    {
+                        lineText = lineText.TrimEnd();
+                    }
+                    else if (caretEndIndex == expectedText.Length)
+                    {
+                        lineText = lineText.TrimStart();
+                    }
+                    else
+                    {
+                        lineText = lineText.Trim();
+                    }
+                }
+
+                var lineTextBeforeCaret = caretStartIndex < lineText.Length
+                    ? lineText.Substring(0, caretStartIndex)
+                    : lineText;
+
+                var lineTextAfterCaret = caretStartIndex < lineText.Length
+                    ? lineText.Substring(caretStartIndex)
+                    : string.Empty;
+
+                Assert.Equal(expectedTextBeforeCaret, lineTextBeforeCaret);
+                Assert.Equal(expectedTextAfterCaret, lineTextAfterCaret);
+                Assert.Equal(expectedTextBeforeCaret.Length + expectedTextAfterCaret.Length, lineText.Length);
+            }
+            else
+            {
+                var lineText = _editor.GetCurrentLineText();
+
+                if (trimWhitespace)
+                {
+                    lineText = lineText.Trim();
+                }
+
+                Assert.Equal(expectedText, lineText);
+            }
+        }
+
+        protected void VerifyTextContains(string expectedText)
+        {
+            var caretStartIndex = expectedText.IndexOf("$$");
+
+            if (caretStartIndex >= 0)
+            {
+                var caretEndIndex = caretStartIndex + "$$".Length;
+
+                var expectedTextBeforeCaret = caretStartIndex < expectedText.Length
```

Comment:
I remember why now. I was just being paranoid because the special cases *are* needed below for `lineTextBeforeCaret` and `lineTextAfterCaret` because there's no guarantee that `caretStartIndex` isn't greater than `lineText.Length`. You're right that `expectedTextBeforeCaret` and `expectedTextAfterCaret` can be changed as you suggest, and I've done so.